# Exercício:

## 01 – Obtenha os resultados e retorne o objeto e código em Python dos seguintes cálculos:
a) 2.0 + 3.0 / 5.0

b) (a + 10.2) - 5.0 * 2.0

c) a / b + 33.8

d) a / (b + c)

e) d * 2

In [47]:
a= 2.0 + 3.0/5.0
b= (a +10.2) - 5.0*2.0
c= a/b + 33.8
d= a/(b+c)
e= d*2
e

0.1385610963075752

## 02 – Obtenha os resultados e retorne o código do objeto em Python dos seguintes cálculos:
a) [2.0, 3.0, 5.0] + [10, 20, 30]

b) lista_a * 2

c) lista_a + lista_b

d) lista_a / (lista_b * lista_c)

In [70]:
lista_a = [2.0,3.0,5.0] + [10,20,30]
lista_a

[2.0, 3.0, 5.0, 10, 20, 30]

In [71]:
lista_b = lista_a * 2
lista_b

[2.0, 3.0, 5.0, 10, 20, 30, 2.0, 3.0, 5.0, 10, 20, 30]

In [72]:
lista_c = lista_a + lista_b
lista_c

[2.0,
 3.0,
 5.0,
 10,
 20,
 30,
 2.0,
 3.0,
 5.0,
 10,
 20,
 30,
 2.0,
 3.0,
 5.0,
 10,
 20,
 30]

In [73]:
lista_d = lista_a / (lista_b * lista_c)
lista_d

TypeError: can't multiply sequence by non-int of type 'list'

## 03 – Usando os objetos simples obtidos no exercício 01 (a, b, c, d, e), instancie uma  lista. Instancie também uma segunda lista contendo a raiz quadrada dos valores obtidos nos objetos simples. Instancie uma terceira lista, concatenando as duas primeiras. Recupere a última lista.

In [52]:
lista = [a, b, c, d, e]
lista2 = [a**0.5, b**0.5, c**0.5, d**0.5, e**0.5]
lista3 = lista + lista2
lista3

[2.6,
 2.799999999999999,
 34.72857142857143,
 0.0692805481537876,
 0.1385610963075752,
 1.61245154965971,
 1.6733200530681507,
 5.893095233285427,
 0.263211983302029,
 0.3722379565648501]

## 04 – Duplique os valores contidos na lista final obtida no exercício anterior e converta para um array numpy. Remodele (reshape) o array, para que contenha 5 linhas e 4 colunas. Recupere o array.

In [ ]:
import numpy as np

In [53]:
lista3 = lista3 * 2
lista3
lista3 = np.array(lista3)
lista3 = lista3.reshape(5,4)
lista3 

array([[ 2.6       ,  2.8       , 34.72857143,  0.06928055],
       [ 0.1385611 ,  1.61245155,  1.67332005,  5.89309523],
       [ 0.26321198,  0.37223796,  2.6       ,  2.8       ],
       [34.72857143,  0.06928055,  0.1385611 ,  1.61245155],
       [ 1.67332005,  5.89309523,  0.26321198,  0.37223796]])

## 05 – Do array numpy obtenha uma lista contendo a média dos valores de cada uma das colunas.

In [54]:
medias = lista3.mean(axis=0)
medias

array([7.88073291, 2.14941306, 7.88073291, 2.14941306])

## 06 – Insira as médias das colunas no final do array do exercício 4.

**Contexto:**
- No exercício 4, você criou um array numpy com 5 linhas e 4 colunas (totalizando 20 valores)
- No exercício 5, você calculou a média de cada uma das 4 colunas (resultando em uma lista com 4 valores)

**Tarefa:**
1. Use o atributo `shape` do array do exercício 4 para calcular o número total de elementos.
2. Insira a lista de médias (do exercício 5) no final de outro array criado
3. Recupere e exiba o array resultante


Obs.:  Atributos não usam () na chamada.

In [55]:
lista3.shape

(5, 4)

In [56]:
lista3 = np.vstack([lista3, medias])
lista3

array([[ 2.6       ,  2.8       , 34.72857143,  0.06928055],
       [ 0.1385611 ,  1.61245155,  1.67332005,  5.89309523],
       [ 0.26321198,  0.37223796,  2.6       ,  2.8       ],
       [34.72857143,  0.06928055,  0.1385611 ,  1.61245155],
       [ 1.67332005,  5.89309523,  0.26321198,  0.37223796],
       [ 7.88073291,  2.14941306,  7.88073291,  2.14941306]])

## 07 – Remodele (reshape) o array do exercício anterior, para que contenha 4 colunas. Converta este objeto da classe array para DataFrame. Arredonde todos os valores para até 3 casas decimais. Recupere a DataFrame.

In [ ]:
import pandas as pd

In [57]:
lista3.reshape(6,4)
lista3 = pd.DataFrame(lista3)
lista3 =np.round(lista3, decimals=3)
lista3 

,0,1,2,3
0,2.600,2.800,34.729,0.069
1,0.139,1.612,1.673,5.893
2,0.263,0.372,2.600,2.800
3,34.729,0.069,0.139,1.612
4,1.673,5.893,0.263,0.372
5,7.881,2.149,7.881,2.149


## 08 – A última instancia desta DataFrame deve conter as médias dos valores contidos por coluna, o que foi determinado nos exercícios anteriores. Instancie um objeto contendo as médias dos valores em cada linha. Insira este objeto como uma nova coluna na DataFrame. Recupere a DataFrame.

In [63]:
medias_df = lista3.mean(axis=1)
medias_df
lista3 = pd.concat([lista3, medias_df], axis=1)
lista3


,0,1,2,3,0
0,2.600,2.800,34.729,0.069,10.04950
1,0.139,1.612,1.673,5.893,2.32925
2,0.263,0.372,2.600,2.800,1.50875
3,34.729,0.069,0.139,1.612,9.13725
4,1.673,5.893,0.263,0.372,2.05025
5,7.881,2.149,7.881,2.149,5.01500


## 09 – Usando o método dos objetos de classe DataFrame, describe(), compare os resultados contidos na última linha e última coluna da DataFrame.

Obs.: O método dos objetos de classe DataFrame, transpose(), inverte linhas por colunas.

In [64]:
ultima_linha = lista3.iloc[-1]
ultima_coluna = lista3.iloc[:, -1]

ultima_linha_t = lista3.iloc[[-1]].transpose()

print("Última linha:")
print(ultima_linha.describe())

print("\nÚltima coluna:")
print(ultima_coluna.describe())

Última linha:
count    5.000
mean     5.015
std      2.866
min      2.149
25%      2.149
50%      5.015
75%      7.881
max      7.881
Name: 5, dtype: float64

Última coluna:
count     6.000000
mean      5.015000
std       3.758615
min       1.508750
25%       2.120000
50%       3.672125
75%       8.106688
max      10.049500
Name: 0, dtype: float64


Embora a última linha e a última coluna apresentem a mesma média (5.015), a última coluna possui maior variabilidade, evidenciada pelo maior desvio padrão e pela maior amplitude entre os valores mínimo e máximo. Além disso, a diferença na quantidade de elementos e na distribuição dos quartis indica que os dados da última coluna são mais dispersos, enquanto a última linha apresenta uma distribuição mais uniforme.

## 10 – Edite o nome das colunas e dos index, para indicar onde estão as médias. Qual a média das médias?

In [65]:
lista3.columns = list(lista3.columns[:-1]) + ["Media_Linhas"]
lista3.index = list(lista3.index[:-1]) + ["Media_Colunas"]
print(lista3)
media_das_medias = lista3.iloc[-1].mean()
print(media_das_medias)

                    0      1       2      3  Media_Linhas
0               2.600  2.800  34.729  0.069      10.04950
1               0.139  1.612   1.673  5.893       2.32925
2               0.263  0.372   2.600  2.800       1.50875
3              34.729  0.069   0.139  1.612       9.13725
4               1.673  5.893   0.263  0.372       2.05025
Media_Colunas   7.881  2.149   7.881  2.149       5.01500
5.015000000000001
